pgxn/neon/neon.c

```diff
@@ -267,6 +268,148 @@ LogicalSlotsMonitorMain(Datum main_arg)
+	 * PGPROC_MAX_CACHED_SUBXIDS=64. - If number of subxacts exceeds
+	 * PGPROC_MAX_CACHED_SUBXIDS then this proc entry is marked as overflowed
+	 * - When GetRunningTransactionData scans procarray to extract information
+	 * about running xids it sets suboverflowed=true if any of entries is
+	 * overflowed. In this case iProcarray contains "cached subxids", their
+	 * number is limited by PGPROC_MAX_CACHED_SUBXIDS=64. If number of
+	 * subxacts exceeds PGPROC_MAX_CACHED_SUBXIDS then this proc entry is
+	 * marked as overflowed - When GetRunningTransactionData scans procarray
+	 * to extract information about running xids it sets suboverflowed=true if
+	 * any of entries is overflowed. In this case nformation about subxides is
+	 * NOT included in RUNNING_XACT record - Maximal size is known xids is
+	 * defined as ((PGPROC_MAX_CACHED_SUBXIDS + 1) * PROCARRAY_MAXPROCS). So
+	 * it means that if none of transactions has more than
+	 * PGPROC_MAX_CACHED_SUBXIDSsubxacts then information about running xacts
+	 * should fit in known xids. As a result we should normally never get "too
+	 * many KnownAssignedXids" with vanilla Postgres. - When starting replica
+	 * see RUNNING_XACTS record with subxid_overflow it switch to state
+	 * STANDBY_SNAPSHOT_PENDING and waits until either it receive non-overflow
+	 * snapshot, either oldestRunningXid is later than the nextXid from the
+	 * initial snapshot.
+	 *
+	 * So if we want to exclude "too many KnownAssignedXids" error then the
+	 * only safe limit for restoring running xacts from CXLOG shoiuld be
+	 * PGPROC_MAX_CACHED_SUBXIDS. But it is too small. Also we take in account
+	 * only subxacts, not top level transactions which number can be grreater
+	 * thsn 64.
+	 *
+	 * Summarizing all above: there are following apporaches and related
+	 * risks: 1. If we restore as much running xacts as it possible
+	 * (TOTAL_MAX_CACHED_SUBXIDS) then there is risk that future transactions
+	 * will trigger e "too many KnownAssignedXids" error which cause crash of
+	 * replica. 2. If we limit number of restored xacts by some small constant
+	 * (PGPROC_MAX_CACHED_SUBXIDS), then miost likely we will have to wait
+	 * until primary send us running-xacts non-overflowed WAL record. It make
+	 * takes quite long time (running xacts records is produced erach 15
+	 * seconds) and if primary is offline it will never happen and sop replica
+	 * is not started. 3. If we restore just some subset of subxacts but
+	 * doesn;t mark snapshot as overflown (and so do not wait for primary),
+	 * then there is a risk of data corruption (wrong hint bits) or incorrect
+	 * query results (brtoken visibility) at replica.
+	 *
+	 * Current solution is to limit subxacts by  TOTAL_MAX_CACHED_SUBXIDS / 2
+	 * and mark snmapshot as overflowed if real number if larger.
+	 */
+	max_xcnt = TOTAL_MAX_CACHED_SUBXIDS / 2;
+
+	xcnt = till > from ? till - from : till - from - FirstNormalTransactionId;
```

Comment:
This isn't correct, when till == from. The result of (till - from - FirstNormaltransactionid) becomes -3.

I bumped into this by accident while testing a completely different hot standby related thing, and was luckily still running the binary from this PR :-). Added a test case to trigger the problem. I'm surprised none of the existing tests covered the case that oldestActiveXid==nextXid, you'd think that that's the common scenario. I guess it's not because of the "bump to next 1024" boundary in the pageserver.

Reply:
I'd suggest adding some parens, or turning this into an if-else statement. The `? :` construct is much less readable when it's not completely trivial.

Reply:
> I'm surprised none of the existing tests covered the case that oldestActiveXid==nextXid, you'd think that that's the common scenario.

Hmmm. I think `oldestActiveXid==nextXid` is invalid condition in any case, isn't it?
oldestActiveXid is xid is some of existed transaction. nextXid is should be greater (taken in account wraparound)  than xid of any existed transaction. If this assumption is violated, then we I trouble in any case.

Reply:
> I'd suggest adding some parens, or turning this into an if-else statement. The ? : construct is much less readable when it's not completely trivial.

I will add parens. IMHO it is a matter of taste. ?: looks more confusing for people not familiar with C.
But somebody prefer it because it is more compact:) From my point of view it will be better to have `if` as expression, so it is possible to write `xcnt = if (till > from)  { till - from } else { till - from - FirstNormalTransactionId }`
as it possible in Rust and in most of other modern C ancestors.

From my point if view advantage of using ?: rather then `if else` is not a matter of compactness but that it more obviously express semantic of operation: assign to variable on or another value.
```
if (...) x = y; else x = z; 
```
is less expressive.